src/util/schnorr.rs

```diff
@@ -17,3 +17,68 @@
 //!
 
 pub use secp256k1::schnorrsig::{PublicKey, KeyPair};
+use secp256k1::{Secp256k1, Verification};
+use hashes::{Hash, HashEngine};
+use util::taproot::{TapBranchHash, TapTweakHash};
+
+/// Untweaked Schnorr public key
+pub type UntweakedPublicKey = PublicKey;
+
+/// Tweaked Schnorr public key
+pub struct TweakedPublicKey(PublicKey);
+
+/// A trait for tweaking Schnorr public keys
+pub trait Tweakable {
+    /// Tweaks an untweaked public key given an untweaked key and optional script tree merkle root.
+    /// 
+    /// This is done by using the equation Q = P + H(P|c)G, where 
+    ///  * Q is the tweaked key
+    ///  * P is the internal key
+    ///  * H is the hash function
+    ///  * c is the commitment data
+    ///  * G is the generator point
+    fn tap_tweak<C: Verification>(&self, secp: Secp256k1<C>, merkle_root: Option<TapBranchHash>) -> TweakedPublicKey;
+
+    /// Directly convert an UntweakedPublicKey to a TweakedPublicKey 
+    fn dangerous_assume_tweaked(self) -> TweakedPublicKey;
+}
+
+impl Tweakable for UntweakedPublicKey {
+    fn tap_tweak<C: Verification>(&self, secp: Secp256k1<C>, merkle_root: Option<TapBranchHash>) -> TweakedPublicKey {
+        // Compute the tweak
+        let mut engine = TapTweakHash::engine();
+        engine.input(&self.serialize());
+        if let Some(hash) = merkle_root {
+            engine.input(&hash);
+        } else {
+            // nothing to hash
+        }
```

Comment:
nit: I usually prefer

```suggestion
        merkle_root.map(|hash| engine.input(hash));
```

as more concise version